### ISSUE TITLE
cleanup: small correctness and doc nits found during #152

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#459]
## Problem

Six small correctness and documentation nits came up during the #152 work. None of them changes a training result today. Each is a fix of a few lines, so they are bundled here as one checklist.

- [ ] **1.** `inferenceBatched` over-allocates its result array: it uses `sizeof(tensor_t)` for an array of `tensor_t *`.
- [ ] **2.** `MSE.h` documents a B=0 fallback that `computeMeanScaleMSE` does not have.
- [ ] **3.** `LossFunction.h` has three stale doc lines: the 4-value `defaultLossConfig` return list, the "implicit batch dim" note, and a pointer to a `docs/CONVENTIONS.md` section that moved.
- [ ] **4.** An orphaned "NARROWED guard" comment sits in `MaxPool1d.c`, `AvgPool1d.c` and `AdaptiveAvgPool1d.c`.
- [ ] **5.** `%lu` is used for `size_t` at 16 `src/` sites, 6 of them in `src/layer/`.
- [ ] **6.** The same rank-1 FLOAT32 tensor builder is copy-pasted into 8 unit-test files.

---

### 1. `inferenceBatched` over-allocates

**Evidence.** `src/userApi/InferenceApi.c:205`:
```c
tensor_t **tensorArr = reserveMemory(batch->size * sizeof(tensor_t));
```
`tensor_t` is four pointers (`src/tensor/include/Tensor.h:28-33`), so the array is 4× the size it needs: 32 instead of 8 bytes per entry on LP64, 16 instead of 4 on ILP32.

**Impact.** Memory safety is fine, since nothing reads past the array. `ODT_MEM_PROFILE` is on in `unit_test_debug`, `unit_test_asan`, `unit_test_ubsan` (which inherits `unit_test_debug`) and `examples_memprofile` (`CMakePresets.json:50`, `:61`, `:66-68`, `:88`). In those builds the profiler counts the extra bytes as live (+24 B per sample on LP64). The only in-tree caller is the unit test (`test/unit/userAPI/UnitTestInferenceApi.c:748`), so no example memory profile is inflated today.

**Fix.** `reserveMemory(batch->size * sizeof(*tensorArr))`.

**Test.** Extend `testInferenceBatchedAddsBatchAxisToNaturalSamples` in `test/unit/userAPI/UnitTestInferenceApi.c`.

First add `StorageApi` to that target's `MORE_LIBS` (`test/unit/userAPI/CMakeLists.txt:71-92`; `UnitTestTrainingLoopApi` already lists it at `:113`). `ODT_MEM_PROFILE` is PUBLIC on `StorageApi` (`src/userApi/CMakeLists.txt:44`), but it only reaches targets that link `StorageApi` directly. `InferenceApi` links its deps PRIVATE and does not list `StorageApi` (`src/userApi/CMakeLists.txt:24-40`), and no other `MORE_LIBS` entry forwards the define. Without this change the `#ifdef` block below compiles out in every preset, and the test passes without checking anything.

Then, after the `freeTensor(outputs[i])` loop:
1. Take `memProfileMark()`.
2. Call `freeReservedMemory(outputs)` (currently `:766`).
3. Take the mark again and assert that the difference equals `batch.size * sizeof(tensor_t *)`. The delta is exact because the counter adds and subtracts the requested byte count, not the header (`src/userApi/StorageApi.c:40-41`, `:52`).

Put this under `#ifdef ODT_MEM_PROFILE`, as `test/unit/userAPI/UnitTestStorageApi.c:48-51` does. Under plain `unit_test` (profiling OFF), `memProfileMark()` returns 0, so the block has to compile out there. Once the CMake change is in, the test is RED today in the profiling presets: 64 instead of 16 for B=2 on LP64. Mutation check: restore `sizeof(tensor_t)`.

### 2. `MSE.h` promises a B=0 fallback that does not exist

**Evidence.** `src/loss_functions/include/MSE.h:17-18`:
> The microbatch dimension (B = dimensions[0]) is treated as 1 if shape is degenerate (B=0).

`src/loss_functions/MSE.c:190-191` divides by it with no check:
```c
size_t microbatch = modelOutput->shape->dimensions[0];
size_t numFeaturesPerSample = calcNumberOfElementsByTensor(modelOutput) / microbatch;
```
`d3d96155` (#135) removed the earlier `microbatch == 0` fallback on purpose: "defensive masking of an unreachable bug case is anti-pattern; division-by-zero is the louder, more debuggable failure mode". The same commit added the header sentence, so the sentence has been wrong since it was written. That commit did not rule out a check, though. It deferred a runtime `dimensions[0] >= 1` assert to #152 / #153, and `docs/conventions/loss.md:72-77` ("Shape assertion (deferred)") still says so. So the silent fallback is gone for good, and the assert is still open work.

**Impact / reachability.** It cannot be reached in-tree. The only non-test caller that can dispatch to `computeMeanScaleMSE` is `trainingEpochDefault` (`src/userApi/training_loop/training_epoch/TrainingEpochDefault.c:51-53`). It passes a `batchViewOf` view, and `batchViewOf` always sets `dims[0] = 1` (`src/userApi/tensor/BatchView.c:18`). `examples/kws_raw/trace_c.c:343-344` uses the same vtable slot, but its loss is hardwired to `CROSS_ENTROPY` (`:321-322`). PR 3b keeps the `trainingEpochDefault` call as it is (spec §6.5). The risk is a reader who trusts the comment.

**Fix.** Change the comment only. Delete the B=0 sentence and state the real contract from `LossFunction.h:32`: `dimensions[0] >= 1` is a precondition that is not asserted yet (see #153). Also say that the in-tree caller passes the `[1, ...]` view of the label, not the model output, as `LossFunction.h:57-61` already does. The runtime assert belongs to #153, not to this issue.

*Side note for #153.* The "louder failure" argument holds on x86, where integer division by zero raises SIGFPE. It does not hold on AArch64 (the Apple-silicon dev host) or on Armv7-M with the default `CCR.DIV_0_TRP = 0`. There, unsigned division by zero returns 0, and the function silently returns `1.0f/0 = +inf`. That is an argument for #153's assert rather than relying on a trap.

**Test.** None, since this is a comment change.

### 3. `LossFunction.h`: three stale doc lines

**Evidence**
- `src/loss_functions/include/LossFunction.h:16`: "Returns {funcType, REDUCTION_MEAN, REDUCTION_MEAN, NULL}". That is four values, but `lossConfig_t` has three fields (`:10-14`) and `defaultLossConfig` sets three (`src/loss_functions/LossFunction.c:12-18`). It has been wrong since `d3d96155` (#135) wrote it right next to the three-field struct that the same commit introduced.
- `:26`: "For B=1 today the [B, ...] dim may be implicit." This contradicts `docs/conventions/loss.md:26-27` ("it is always explicit"). It also no longer matches the loop: since #152 PR3a the loop makes the leading 1 explicit with `batchViewOf` (`TrainingBatchDefault.c:20-21`, `TrainingEpochDefault.c:51-53`).
- `:33-34`: "see docs/CONVENTIONS.md §\"Loss API: microbatch contracts\"". `docs/CONVENTIONS.md` has no such section (its headings are at `:1`, `:11`, `:19`, `:37`). `a3bf34c8` moved it to `docs/conventions/loss.md:3`.

**Fix**
- `:16`: `{funcType, REDUCTION_MEAN, NULL}`.
- `:26`: "Always explicit: the training loop adds the leading 1 via `batchViewOf` (#152 PR3a); see docs/conventions/loss.md, 'Microbatch shape'."
- `:34`: point at docs/conventions/loss.md, "Loss API: microbatch contracts".
- Optional, while there: `:58` says "The only caller (trainingEpochDefault)". `examples/kws_raw/trace_c.c:343` also calls `computeMeanScale` (CE only, also with a `batchViewOf` label view), so "the in-tree callers" is more accurate.

**Test.** None, since this is a comment change.

### 4. Orphaned "NARROWED guard" comment in the three pools

**Evidence.** The same block appears in three places: `src/layer/MaxPool1d.c:352-358`, `src/layer/AvgPool1d.c:308-313` and `src/layer/AdaptiveAvgPool1d.c:303-308`. MaxPool's copy has one extra clause: "(the argmax indices recorded by the forward carry all the routing)".
```c
/* BFP epic PR4 (R-P4): the ARITH_BFP arm below IS the native pooling path, so
 * this guard is NARROWED to the two arms that raw-view ->data in their own
 * storage format — ...
```
A blank line and then `*BackwardFloat` follow it, so "this guard" no longer points at anything.

History:
- `599dfcaa`, `d273cab2` and `72f35a68` gave these blocks their current wording, as the doc comments of a per-file `static void requireNoBfpWire(...)` in MaxPool1d, AvgPool1d and AdaptiveAvgPool1d.
- `2894d8c4` promoted that helper to the shared `bfpRequireNoBfpWire` (`src/arithmetic/include/BfpKernelSupport.h:151`) and deleted the three per-file functions, but left their comments behind.

The guard calls the comment describes are in each backward dispatcher (e.g. `MaxPool1d.c:563-564` and `:568-569`). The FLOAT32 arm there already has its own rationale comment (`MaxPool1d.c:560-562`, `AvgPool1d.c:481-483`, `AdaptiveAvgPool1d.c:467-469`). The block's content is still accurate; it is just detached and mostly redundant.

**Fix.** Delete the block in all three files. The only sentence the arm comments do not already cover is "forwardInput is NOT guarded: this layer never dereferences it". Move it above the dispatcher if it is worth keeping.

`Relu.c:17-22` and `Dropout.c:43-47` also have "NARROWED" comments. Those are still attached to a live local `requireBfpWire` and are fine.

**Test.** None, since this is a comment change.

### 5. `%lu` for `size_t`

**Evidence.** There are six rank-guard messages in `src/layer/`, all in `*CalcOutputShape`: `Linear.c:566`, `Conv1d.c:1032`, `Conv1dTransposed.c:1069`, `AvgPool1d.c:533`, `MaxPool1d.c:612` and `AdaptiveAvgPool1d.c:519`. For example:
```c
PRINT_ERROR("Linear layer expects 2D input, got %luD\n", inputShape->numberOfDimensions);
```
Ten more sites print `size_t` with `%lu`: `src/arithmetic/Arithmetic.c:18,33,45`, `src/arithmetic/Matmul.c:56,144,340,519` and `src/tensor/Tensor.c:385,389,394`. Everything else in `src/` uses `%zu` (45 files).

Some of these strings also end in `\n`: `Linear.c:566` and `Arithmetic.c:33` (both `PRINT_ERROR`), plus the `PRINT_DEBUG` sites in `Arithmetic.c:45` and `Matmul.c`. Those macros already append `"\033[0m\n"` (`src/common/include/Common.h:45`, `:27`). The result is an extra line break, with the colour reset on its own line.

**Impact.** On the ILP32 MCU target `size_t` is `unsigned int`, so `%lu` is formally undefined behaviour. The printed values are still right in practice because both types are 32-bit there. CI does not catch it: the `arm_cross` preset builds with `-Werror` but without `-Wall`/`-Wformat` (`CMakePresets.json:92-100`). I checked this locally with arm-none-eabi-gcc 10.3 (`-mcpu=cortex-m4 -mthumb`):
- `-Werror` alone compiles with no warning.
- `-Wformat` warns: `format '%lu' expects argument of type 'long unsigned int', but argument 2 has type 'size_t' {aka 'unsigned int'}`.

**Fix.** Change `%lu` to `%zu` at all 16 sites. Drop the trailing `\n` at the `PRINT_ERROR`/`PRINT_DEBUG` sites; the raw `printf` calls in `Tensor.c` keep theirs.

**Test.** A unit test cannot check a format specifier. Two options:
- (a) A CI grep gate that fails if `git grep -n '%lu' -- src/` finds anything, like the `optimizer-step-entry` gate.
- (b) Add `-Wformat` to the `arm_cross` flags. This catches the whole class of mismatch. I have not tried it on the current tree, so check what else it flags first.

### 6. Rank-1 FLOAT32 test builder is copy-pasted

**Evidence.** Every copy has the same body: `reserveMemory` for dims, order and shape, then `initTensor(..., quantizationInitFloat(), NULL)`, then `tensorFillFromFloatBuffer`.

| File | Helper |
|---|---|
| `test/unit/userAPI/UnitTestTrainingLoopApi.c:67` | `buildFloatTensor1D(n, src)` |
| `test/unit/userAPI/UnitTestMultiLayerTraining.c:731` | `buildFloatTensor1D(n, values)` |
| `test/unit/userAPI/UnitTestGroupNormIntegration.c:108` | `build1DFloat(n, data)` |
| `test/unit/userAPI/UnitTestInferenceApi.c:718` | `buildFloatTensor1DInf(n, values)` |
| `test/unit/common/UnitTestOdtHook.c:122` | `makeVec2(a, b)` (n fixed at 2) |
| `test/unit/optimizer/UnitTestSgd.c:1803` | `buildFloatTensor1D(src, n)` (args swapped; `src == NULL` gives a zero-filled tensor) |
| `test/unit/optimizer/UnitTestAdamW.c:103` | `makeFloatTensor1D(src, n)` (same as Sgd) |
| `test/unit/layer/UnitTestRelu.c:376` | `buildFloatTensor1D(n)` (never filled) |

**Fix.** Add one header-only builder under `test/unit/support/include/`, e.g. `TensorBuilders.h` with `static inline tensor_t *buildFloatTensor1D(size_t n, const float *src)`, where `src == NULL` gives a zero-filled tensor. That one signature covers all eight copies.
- The body is the Rule 1 pattern from `docs/conventions/testing.md:145-158`: `initTensor` plus the copying `tensorFillFromFloatBuffer`.
- `testing.md:164-167` defers a cross-file helper "until 3+ test files repeat the same construction". Eight files do. Update that paragraph (and Rule 1) to point at the new header.
- Register it as an INTERFACE library the way `test/unit/support/CMakeLists.txt` does for `odt_test_death` and `odt_test_layer_builders`, and list it in each consumer's `MORE_LIBS`.

Do not migrate `UnitTestMSE.c:196` or `UnitTestCrossEntropy.c:416`. Those helpers are also named `buildFloatTensor1D`, but they build a rank-2 `[1, n]` tensor, so they are misnamed rather than duplicates. If either file ever includes the new header, the names clash and one must be renamed (e.g. to `buildFloatRow`).

Optional: the FLOAT32 2-D builder is copied the same way (`buildFloatTensor2D` / `build2DFloat` / `buildFloatTensor2DInf` in 7 `test/unit/userAPI/` files) and could go into the same header.

**Test.** This is a pure refactor; the migrated suites staying green is the check.

## Found during

- The #152 design spec §8 lists item 1.
- The #456 / #152 PR 3b work surfaced items 2–6.
- Checking this issue against the code turned up the rest: the `LossFunction.h:16` and `:33-34` lines, the second example caller of `computeMeanScale`, the two extra pool files in item 4, the ten `%lu` sites outside `src/layer/`, and the three extra builder copies.

Refs #152
